// @php-wasm/web-service-worker/index.js
import { isURLScoped as R, setURLScope as C, getURLScope as y } from "@php-wasm/scopes";
const E = 25e3;
let b = 0;
function H(e, t, ...o) {
  const s = w();
  return e.postMessage(
    {
      ...t,
      requestId: s
    },
    ...o
  ), s;
}
function w() {
  return ++b;
}
function L(e, t, o = E) {
  return new Promise((s, n) => {
    const r = (i) => {
      i.data.type === "response" && i.data.requestId === t && (e.removeEventListener("message", r), clearTimeout(a), s(i.data.response));
    }, a = setTimeout(() => {
      n(new Error("Request timed out")), e.removeEventListener("message", r);
    }, o);
    e.addEventListener("message", r);
  });
}
function q(e, t) {
  return {
    type: "response",
    requestId: e,
    response: t
  };
}
async function I(e) {
  let t = new URL(e.request.url);
  if (!R(t))
    try {
      const c = new URL(e.request.referrer);
      t = C(t, y(c));
    } catch {
    }
  const o = e.request.headers.get("content-type"), s = e.request.method === "POST" ? new Uint8Array(await e.request.clone().arrayBuffer()) : void 0, n = {};
  for (const c of e.request.headers.entries())
    n[c[0]] = c[1];
  let r;
  try {
    const c = {
      method: "request",
      args: [
        {
          body: s,
          url: t.toString(),
          method: e.request.method,
          headers: {
            ...n,
            Host: t.host,
            // Safari and Firefox don't make the User-Agent header
            // available in the fetch event. Let's add it manually:
            "User-agent": self.navigator.userAgent,
            "Content-type": o
          }
        }
      ]
    }, d = y(t);
    if (d === null)
      throw new Error(
        `The URL ${t.toString()} is not scoped. This should not happen.`
      );
    const h = await U(c, d);
    if (r = await L(self, h), delete r.headers["x-frame-options"], r.headers["content-security-policy"]) {
      const l = r.headers["content-security-policy"].map(
        (u) => T(
          "frame-ancestors",
          u
        )
      ).filter((u) => u.trim().length > 0);
      l.length > 0 ? r.headers["content-security-policy"] = l : delete r.headers["content-security-policy"];
    }
  } catch (c) {
    throw console.error(c, { url: t.toString() }), c;
  }
  if (r.httpStatusCode >= 300 && r.httpStatusCode <= 399 && r.headers.location)
    return Response.redirect(
      new URL(r.headers.location[0], t.toString()),
      r.httpStatusCode
    );
  const i = [101, 103, 204, 205, 304].includes(
    r.httpStatusCode
  ) ? null : r.bytes;
  return new Response(i, {
    headers: r.headers,
    status: r.httpStatusCode
  });
}
async function U(e, t) {
  const o = w();
  for (const s of await self.clients.matchAll({
    // Sometimes the client that triggered the current fetch()
    // event is considered uncontrolled in Google Chrome. This
    // only happens on the first few fetches() after the initial
    // registration of the service worker.
    includeUncontrolled: !0
  }))
    s.postMessage({
      ...e,
      /**
       * Attach the scope with a URL starting with `/scope:` to this message.
       *
       * We need this mechanics because this worker broadcasts
       * events to all the listeners across all browser tabs. Scopes
       * helps WASM workers ignore requests meant for other WASM workers.
       */
      scope: t,
      requestId: o
    });
  return o;
}
async function f(e, t) {
  let o;
  return ["GET", "HEAD"].includes(e.method) || "body" in t ? o = void 0 : !e.bodyUsed && e.body ? o = e.body : o = await e.arrayBuffer(), new Request(t.url || e.url, {
    body: o,
    method: e.method,
    headers: e.headers,
    referrer: e.referrer,
    referrerPolicy: e.referrerPolicy,
    mode: e.mode === "navigate" ? "same-origin" : e.mode,
    credentials: e.credentials,
    cache: e.cache,
    redirect: e.redirect,
    integrity: e.integrity,
    ...o && { duplex: "half" },
    ...t
  });
}
async function S(e) {
  if (!e.body)
    return [e, e];
  const [t, o] = e.body.tee();
  return [
    await f(e, { body: t, duplex: "half" }),
    await f(e, { body: o, duplex: "half" })
  ];
}
function O(e) {
  const t = {};
  return e.headers.forEach((o, s) => {
    t[s] = o;
  }), t;
}
function T(e, t) {
  const o = /^[\u{9}\u{A}\u{C}\u{D}\u{20}]+/u, s = /[\u{9}\u{A}\u{C}\u{D}\u{20}]+$/u, n = /[\u{9}\u{A}\u{C}\u{D}\u{20}]/u;
  return t.split(";").filter((r) => {
    const a = r.replace(o, "").replace(s, ""), [i] = a.split(
      n,
      // The directive name is the first token.
      1
    );
    return i.toLowerCase() !== e.toLowerCase();
  }).join(";");
}
class x extends Error {
  constructor(t, o, s) {
    super(
      `Could not fetch ${t} – your network appears to be blocking this request (HTTP ${o}). This often happens on school, university, or corporate networks. Try switching to a different network or using a VPN.`
    ), this.name = "FirewallInterferenceError", this.url = t, this.status = o, this.statusText = s;
  }
}
const P = "X-Playground-Cors-Proxy";
async function D(e, t, o, s) {
  var h;
  let n = typeof e == "string" ? new Request(e, t) : e;
  const r = s ? new URL(s) : null;
  let a = r ? new URL(n.url, r) : new URL(n.url);
  if (a.hostname === "localhost" || a.hostname === "127.0.0.1" || a.hostname === "[::1]" || a.hostname === "::1")
    return await fetch(n);
  if (a.protocol === "http:") {
    a.protocol = "https:";
    const l = a.toString();
    n = await f(n, { url: l }), a = new URL(l);
  }
  if (!o)
    return await fetch(n);
  if (r && a.protocol === r.protocol && a.hostname === r.hostname && a.port === r.port && a.pathname.startsWith(r.pathname))
    return await fetch(n);
  const [c, d] = await S(n);
  try {
    return await fetch(c);
  } catch {
    const u = ((h = new Headers(d.headers).get("x-cors-proxy-allowed-request-headers")) == null ? void 0 : h.split(",")) || [], m = u.includes("authorization") || u.includes("cookie"), g = await f(d, {
      url: `${o}${n.url}`,
      ...m && { credentials: "include" }
    }), p = await fetch(g, t);
    if (!p.headers.has(P))
      throw new x(
        n.url,
        p.status,
        p.statusText
      );
    return p;
  }
}
export {
  x as FirewallInterferenceError,
  L as awaitReply,
  U as broadcastMessageExpectReply,
  f as cloneRequest,
  I as convertFetchEventToPHPRequest,
  D as fetchWithCorsProxy,
  w as getNextRequestId,
  O as getRequestHeaders,
  H as postMessageExpectReply,
  T as removeContentSecurityPolicyDirective,
  q as responseTo,
  S as teeRequest
};
