// @wp-playground/cli/run-cli-Cv4FHhj_.js
import { logger as P, LogSeverity as F, errorLogPath as Z } from "@php-wasm/logger";
import { consumeAPI as ue, ProcessIdAllocator as xe, SupportedPHPVersions as R, printDebugDetails as $e, HttpCookieStore as Ee, FileLockManagerInMemory as Te, StreamedPHPResponse as J, createObjectPoolProxy as Ie, exposeAPI as ke, PHPResponse as Me, exposeSyncAPI as Be } from "@php-wasm/universal";
import { resolveRemoteBlueprint as Ae, resolveRuntimeConfiguration as K, compileBlueprintV1 as de, isBlueprintBundle as Ce, runBlueprintV1Steps as ee } from "@wp-playground/blueprints";
import { zipDirectory as De, RecommendedPHPVersion as M } from "@wp-playground/common";
import c, { existsSync as Y, rmdirSync as Re, mkdirSync as j, readdirSync as We } from "fs";
import { MessageChannel as pe, Worker as te } from "worker_threads";
import { createNodeFsMountHandler as Le } from "@php-wasm/node";
import d, { basename as D, join as ce } from "path";
import { exec as me } from "child_process";
import { promisify as Ue } from "util";
import fe from "express";
import { Readable as He } from "stream";
import { pipeline as Fe } from "stream/promises";
import Ne from "yargs";
import { NodeJsFilesystem as Oe, OverlayFilesystem as _e, InMemoryFilesystem as je, ZipFilesystem as Ve } from "@wp-playground/storage";
import { EmscriptenDownloadMonitor as Ye, ProgressTracker as qe } from "@php-wasm/progress";
import { resolveWordPressRelease as ze } from "@wp-playground/wordpress";
import $ from "fs-extra";
import { createRequire as Ge } from "module";
import q from "os";
import { startBridge as Qe } from "@php-wasm/xdebug-bridge";
import { dir as Xe, setGracefulCleanup as Ze } from "tmp-promise";
import Je from "ps-man";
import { removeTempDirSymlink as Ke, createTempDirSymlink as oe, makeXdebugConfig as et, clearXdebugIDEConfig as tt, addXdebugIDEConfig as ot } from "@php-wasm/cli-util";
import { createHash as rt } from "crypto";
import { PHPMYADMIN_INSTALL_PATH as re, getPhpMyAdminInstallSteps as st, PHPMYADMIN_ENTRY_PATH as it } from "@wp-playground/tools";
import { jspi as se } from "wasm-feature-detect";
function N(e) {
  const t = [];
  for (const o of e) {
    const r = o.split(":");
    if (r.length !== 2)
      throw new Error(`Invalid mount format: ${o}.
				Expected format: /host/path:/vfs/path.
				If your path contains a colon, e.g. C:\\myplugin, use the --mount-dir option instead.
				Example: --mount-dir C:\\my-plugin /wordpress/wp-content/plugins/my-plugin`);
    const [s, i] = r;
    if (!Y(s))
      throw new Error(`Host path does not exist: ${s}`);
    t.push({ hostPath: s, vfsPath: i });
  }
  return t;
}
function ie(e) {
  if (e.length % 2 !== 0)
    throw new Error("Invalid mount format. Expected: /host/path /vfs/path");
  const t = [];
  for (let o = 0; o < e.length; o += 2) {
    const r = e[o], s = e[o + 1];
    if (!Y(r))
      throw new Error(`Host path does not exist: ${r}`);
    t.push({
      hostPath: d.resolve(process.cwd(), r),
      vfsPath: s
    });
  }
  return t;
}
async function yo(e, t) {
  for (const o of t)
    await e.mount(
      o.vfsPath,
      Le(o.hostPath)
    );
}
const ne = {
  step: "runPHP",
  code: {
    filename: "activate-theme.php",
    // @TODO: Remove DOCROOT check after moving totally to Blueprints v2.
    content: `<?php
			$docroot = getenv('DOCROOT') ? getenv('DOCROOT') : '/wordpress';
			require_once "$docroot/wp-load.php";
			$theme = wp_get_theme();
			if (!$theme->exists()) {
				$themes = wp_get_themes();
				if (count($themes) > 0) {
					$themeName = array_keys($themes)[0];
					switch_theme($themeName);
				}
			}
		`
  }
};
function he(e) {
  const t = e.autoMount, o = [...e.mount || []], r = [...e["mount-before-install"] || []], s = {
    ...e,
    mount: o,
    "mount-before-install": r,
    "additional-blueprint-steps": [
      ...e["additional-blueprint-steps"] || []
    ]
  };
  if (ut(t)) {
    const n = `/wordpress/wp-content/plugins/${D(t)}`;
    o.push({
      hostPath: t,
      vfsPath: n,
      autoMounted: !0
    }), s["additional-blueprint-steps"].push({
      step: "activatePlugin",
      pluginPath: `/wordpress/wp-content/plugins/${D(t)}`
    });
  } else if (lt(t)) {
    const i = D(t), n = `/wordpress/wp-content/themes/${i}`;
    o.push({
      hostPath: t,
      vfsPath: n,
      autoMounted: !0
    }), s["additional-blueprint-steps"].push(
      e["experimental-blueprints-v2-runner"] ? {
        step: "activateTheme",
        themeDirectoryName: i
      } : {
        step: "activateTheme",
        themeFolderName: i
      }
    );
  } else if (at(t)) {
    const i = c.readdirSync(t);
    for (const n of i)
      n !== "index.php" && o.push({
        hostPath: `${t}/${n}`,
        vfsPath: `/wordpress/wp-content/${n}`,
        autoMounted: !0
      });
    s["additional-blueprint-steps"].push(ne);
  } else nt(t) && (r.push({
    hostPath: t,
    vfsPath: "/wordpress",
    autoMounted: !0
  }), s.mode = "apply-to-existing-site", s["additional-blueprint-steps"].push(ne), s.wordpressInstallMode || (s.wordpressInstallMode = "install-from-existing-files-if-needed"));
  return s;
}
function nt(e) {
  const t = c.readdirSync(e);
  return t.includes("wp-admin") && t.includes("wp-includes") && t.includes("wp-content");
}
function at(e) {
  const t = c.readdirSync(e);
  return t.includes("themes") || t.includes("plugins") || t.includes("mu-plugins") || t.includes("uploads");
}
function lt(e) {
  if (!c.readdirSync(e).includes("style.css"))
    return !1;
  const o = c.readFileSync(ce(e, "style.css"), "utf8");
  return !!/^(?:[ \t]*<\?php)?[ \t/*#@]*Theme Name:(.*)$/im.exec(o);
}
function ut(e) {
  const t = c.readdirSync(e), o = /^(?:[ \t]*<\?php)?[ \t/*#@]*Plugin Name:(.*)$/im;
  return !!t.filter((s) => s.endsWith(".php")).find((s) => {
    const i = c.readFileSync(ce(e, s), "utf8");
    return !!o.exec(i);
  });
}
function dt(e) {
  if (e.length % 2 !== 0)
    throw new Error(
      "Invalid constant definition format. Expected pairs of NAME value"
    );
  const t = {};
  for (let o = 0; o < e.length; o += 2) {
    const r = e[o], s = e[o + 1];
    if (!r || !r.trim())
      throw new Error("Constant name cannot be empty");
    t[r.trim()] = s;
  }
  return t;
}
function pt(e) {
  if (e.length % 2 !== 0)
    throw new Error(
      "Invalid boolean constant definition format. Expected pairs of NAME value"
    );
  const t = {};
  for (let o = 0; o < e.length; o += 2) {
    const r = e[o], s = e[o + 1].trim().toLowerCase();
    if (!r || !r.trim())
      throw new Error("Constant name cannot be empty");
    if (s === "true" || s === "1")
      t[r.trim()] = !0;
    else if (s === "false" || s === "0")
      t[r.trim()] = !1;
    else
      throw new Error(
        `Invalid boolean value for constant "${r}": "${s}". Must be "true", "false", "1", or "0".`
      );
  }
  return t;
}
function ct(e) {
  if (e.length % 2 !== 0)
    throw new Error(
      "Invalid number constant definition format. Expected pairs of NAME value"
    );
  const t = {};
  for (let o = 0; o < e.length; o += 2) {
    const r = e[o], s = e[o + 1].trim();
    if (!r || !r.trim())
      throw new Error("Constant name cannot be empty");
    const i = Number(s);
    if (isNaN(i))
      throw new Error(
        `Invalid number value for constant "${r}": "${s}". Must be a valid number.`
      );
    t[r.trim()] = i;
  }
  return t;
}
function mt(e = {}, t = {}, o = {}) {
  const r = {}, s = /* @__PURE__ */ new Set(), i = (n, l) => {
    for (const u in n) {
      if (s.has(u))
        throw new Error(
          `Constant "${u}" is defined multiple times across different --define-${l} flags`
        );
      s.add(u), r[u] = n[u];
    }
  };
  return i(e, "string"), i(t, "bool"), i(o, "number"), r;
}
function we(e) {
  return mt(
    e.define,
    e["define-bool"],
    e["define-number"]
  );
}
const ft = Ue(me);
function ht(e) {
  return new Promise((t) => {
    if (e === 0) return t(!1);
    const o = fe().listen(e);
    o.once("listening", () => o.close(() => t(!1))), o.once(
      "error",
      (r) => t(r.code === "EADDRINUSE")
    );
  });
}
async function wt(e) {
  const t = fe(), o = await new Promise((n, l) => {
    const u = t.listen(e.port, () => {
      const w = u.address();
      w === null || typeof w == "string" ? l(new Error("Server address is not available")) : n(u);
    }).once("error", l);
  });
  t.use("/", async (n, l) => {
    try {
      const u = {
        url: n.url,
        headers: Pt(n),
        method: n.method,
        body: await bt(n)
      }, w = await e.handleRequest(u);
      await yt(w, l);
    } catch (u) {
      P.error(u), l.headersSent || (l.statusCode = 500, l.end("Internal Server Error"));
    }
  });
  const s = o.address().port, i = process.env.CODESPACE_NAME;
  return i && gt(s, i), await e.onBind(o, s);
}
async function yt(e, t) {
  const [o, r] = await Promise.all([
    e.headers,
    e.httpStatusCode
  ]);
  t.statusCode = r;
  for (const i in o)
    t.setHeader(i, o[i]);
  const s = He.fromWeb(e.stdout);
  await Fe(s, t);
}
const bt = async (e) => await new Promise((t) => {
  const o = [];
  e.on("data", (r) => {
    o.push(r);
  }), e.on("end", () => {
    t(new Uint8Array(Buffer.concat(o)));
  });
});
async function gt(e, t) {
  P.log(`Publishing port ${e}...`);
  const o = `gh codespace ports visibility ${e}:public -c ${t}`;
  for (let r = 0; r < 10; r++)
    try {
      await ft(o);
      return;
    } catch {
      await new Promise((s) => setTimeout(s, 2e3));
    }
}
const Pt = (e) => {
  const t = {};
  if (e.rawHeaders && e.rawHeaders.length)
    for (let o = 0; o < e.rawHeaders.length; o += 2)
      t[e.rawHeaders[o].toLowerCase()] = e.rawHeaders[o + 1];
  return t;
};
function vt(e) {
  return /^latest$|^trunk$|^nightly$|^(?:(\d+)\.(\d+)(?:\.(\d+))?)((?:-beta(?:\d+)?)|(?:-RC(?:\d+)?))?$/.test(e);
}
async function St({
  sourceString: e,
  blueprintMayReadAdjacentFiles: t
}) {
  if (!e)
    return;
  if (e.startsWith("http://") || e.startsWith("https://"))
    return await Ae(e);
  let o = d.resolve(process.cwd(), e);
  if (!c.existsSync(o))
    throw new Error(`Blueprint file does not exist: ${o}`);
  const r = c.statSync(o);
  if (r.isDirectory() && (o = d.join(o, "blueprint.json")), !r.isFile() && r.isSymbolicLink())
    throw new Error(
      `Blueprint path is neither a file nor a directory: ${o}`
    );
  const s = d.extname(o);
  switch (s) {
    case ".zip":
      return Ve.fromArrayBuffer(
        c.readFileSync(o).buffer
      );
    case ".json": {
      const i = c.readFileSync(o, "utf-8");
      try {
        JSON.parse(i);
      } catch {
        throw new Error(
          `Blueprint file at ${o} is not a valid JSON file`
        );
      }
      const n = d.dirname(o), l = new Oe(n);
      return new _e([
        new je({
          "blueprint.json": i
        }),
        /**
         * Wrap the NodeJS filesystem to prevent access to local files
         * unless the user explicitly allowed it.
         */
        {
          read(u) {
            if (!t)
              throw new Error(
                `Error: Blueprint contained tried to read a local file at path "${u}" (via a resource of type "bundled"). Playground restricts access to local resources by default as a security measure. 

You can allow this Blueprint to read files from the same parent directory by explicitly adding the --blueprint-may-read-adjacent-files option to your command.`
              );
            return l.read(u);
          }
        }
      ]);
    }
    default:
      throw new Error(
        `Unsupported blueprint file extension: ${s}. Only .zip and .json files are supported.`
      );
  }
}
class xt {
  constructor(t, o) {
    this.args = t, this.siteUrl = o.siteUrl, this.phpVersion = t.php, this.cliOutput = o.cliOutput;
  }
  getWorkerType() {
    return "v2";
  }
  async bootWordPress(t, o) {
    const r = {
      command: this.args.command,
      siteUrl: this.siteUrl,
      blueprint: this.args.blueprint,
      workerPostInstallMountsPort: o
    };
    return await t.bootWordPress(r, o), t;
  }
  async bootRequestHandler({
    worker: t,
    fileLockManagerPort: o,
    nativeInternalDirPath: r
  }) {
    const s = ue(t.phpPort);
    await s.useFileLockManager(o);
    const i = {
      ...this.args,
      phpVersion: this.phpVersion,
      siteUrl: this.siteUrl,
      processId: t.processId,
      trace: this.args.verbosity === "debug",
      withIntl: this.args.intl,
      withRedis: this.args.redis,
      withMemcached: this.args.memcached,
      withXdebug: !!this.args.xdebug,
      nativeInternalDirPath: r,
      mountsBeforeWpInstall: this.args["mount-before-install"] || [],
      mountsAfterWpInstall: this.args.mount || [],
      constants: we(this.args)
    };
    return await s.bootWorker(i), s;
  }
}
const V = d.join(q.homedir(), ".wordpress-playground");
async function $t() {
  const e = typeof __dirname < "u" ? __dirname : import.meta.dirname;
  let t = d.join(e, "sqlite-database-integration.zip");
  if (!$.existsSync(t)) {
    const o = Ge(import.meta.url), r = d.dirname(
      o.resolve("@wp-playground/wordpress-builds/package.json")
    );
    t = d.join(
      r,
      "src",
      "sqlite-database-integration",
      "sqlite-database-integration-trunk.zip"
    );
  }
  return new File([await $.readFile(t)], d.basename(t));
}
async function Et(e, t, o) {
  const r = d.join(V, t);
  return $.existsSync(r) || ($.ensureDirSync(V), await Tt(e, r, o)), ye(r);
}
async function Tt(e, t, o) {
  const s = (await o.monitorFetch(fetch(e))).body.getReader(), i = `${t}.partial`, n = $.createWriteStream(i);
  for (; ; ) {
    const { done: l, value: u } = await s.read();
    if (u && n.write(u), l)
      break;
  }
  n.close(), n.closed || await new Promise((l, u) => {
    n.on("finish", () => {
      $.renameSync(i, t), l(null);
    }), n.on("error", (w) => {
      $.removeSync(i), u(w);
    });
  });
}
function ye(e, t) {
  return new File([$.readFileSync(e)], D(e));
}
class It {
  constructor(t, o) {
    this.args = t, this.siteUrl = o.siteUrl, this.cliOutput = o.cliOutput;
  }
  getWorkerType() {
    return "v1";
  }
  async bootWordPress(t, o) {
    let r, s, i;
    const n = new Ye();
    if (this.args.wordpressInstallMode === "download-and-install") {
      let w = !1;
      n.addEventListener("progress", (S) => {
        if (w)
          return;
        const { loaded: p, total: a } = S.detail, m = Math.floor(
          Math.min(100, 100 * p / a)
        );
        w = m === 100, this.cliOutput.updateProgress(
          "Downloading WordPress",
          m
        );
      }), r = await ze(this.args.wp), i = d.join(
        V,
        `prebuilt-wp-content-for-wp-${r.version}.zip`
      ), s = c.existsSync(i) ? ye(i) : await Et(
        r.releaseUrl,
        `${r.version}.zip`,
        n
      ), P.debug(
        `Resolved WordPress release URL: ${r?.releaseUrl}`
      );
    }
    let l;
    this.args.skipSqliteSetup ? (P.debug("Skipping SQLite integration plugin setup..."), l = void 0) : (this.cliOutput.updateProgress("Preparing SQLite database"), l = await $t()), this.cliOutput.updateProgress("Booting WordPress");
    const u = await K(
      this.getEffectiveBlueprint()
    );
    return await t.bootWordPress(
      {
        wpVersion: u.wpVersion,
        siteUrl: this.siteUrl,
        wordpressInstallMode: this.args.wordpressInstallMode || "download-and-install",
        wordPressZip: s && await s.arrayBuffer(),
        sqliteIntegrationPluginZip: await l?.arrayBuffer(),
        constants: we(this.args)
      },
      o
    ), i && !this.args["mount-before-install"] && !c.existsSync(i) && (this.cliOutput.updateProgress("Caching WordPress for next boot"), c.writeFileSync(
      i,
      // Comlink proxy is not assignable to UniversalPHP but
      // proxies all method calls transparently at runtime.
      await De(
        t,
        "/wordpress"
      )
    )), t;
  }
  async bootRequestHandler({
    worker: t,
    fileLockManagerPort: o,
    nativeInternalDirPath: r
  }) {
    const s = ue(
      t.phpPort
    );
    await s.isConnected();
    const i = await K(
      this.getEffectiveBlueprint()
    );
    return await s.useFileLockManager(o), await s.bootRequestHandler({
      phpVersion: i.phpVersion,
      siteUrl: this.siteUrl,
      mountsBeforeWpInstall: this.args["mount-before-install"] || [],
      mountsAfterWpInstall: this.args.mount || [],
      processId: t.processId,
      followSymlinks: this.args.followSymlinks === !0,
      trace: this.args.experimentalTrace === !0,
      withIntl: this.args.intl,
      withRedis: this.args.redis,
      withMemcached: this.args.memcached,
      withXdebug: !!this.args.xdebug,
      nativeInternalDirPath: r,
      pathAliases: this.args.pathAliases
    }), await s.isReady(), s;
  }
  async compileInputBlueprint(t) {
    const o = this.getEffectiveBlueprint(), r = new qe();
    let s = "", i = !1;
    return r.addEventListener("progress", (n) => {
      if (i)
        return;
      i = n.detail.progress === 100;
      const l = Math.floor(n.detail.progress);
      s = n.detail.caption || s || "Running Blueprint", this.cliOutput.updateProgress(s.trim(), l);
    }), await de(o, {
      progress: r,
      additionalSteps: t
    });
  }
  getEffectiveBlueprint() {
    const t = this.args.blueprint;
    return Ce(t) ? t : {
      login: this.args.login,
      ...t || {},
      preferredVersions: {
        php: this.args.php ?? t?.preferredVersions?.php ?? M,
        wp: this.args.wp ?? t?.preferredVersions?.wp ?? "latest",
        ...t?.preferredVersions || {}
      }
    };
  }
}
async function kt(e, t = !0) {
  const r = `${d.basename(process.argv0)}${e}${process.pid}-`, s = await Xe({
    prefix: r,
    /*
     * Allow recursive cleanup on process exit.
     *
     * NOTE: I worried about whether this cleanup would follow symlinks
     * and delete target files instead of unlinking the symlink,
     * but this feature uses rimraf under the hood which respects symlinks:
     * https://github.com/raszi/node-tmp/blob/3d2fe387f3f91b13830b9182faa02c3231ea8258/lib/tmp.js#L318
     */
    unsafeCleanup: !0
  });
  return t && Ze(), s;
}
async function Mt(e, t, o) {
  const s = (await Bt(
    e,
    t,
    o
  )).map(
    (i) => new Promise((n) => {
      c.rm(i, { recursive: !0 }, (l) => {
        l ? P.warn(
          `Failed to delete stale Playground temp dir: ${i}`,
          l
        ) : P.info(
          `Deleted stale Playground temp dir: ${i}`
        ), n();
      });
    })
  );
  await Promise.all(s);
}
async function Bt(e, t, o) {
  try {
    const r = c.readdirSync(o).map((i) => d.join(o, i)), s = [];
    for (const i of r)
      await At(
        e,
        t,
        i
      ) && s.push(i);
    return s;
  } catch (r) {
    return P.warn(`Failed to find stale Playground temp dirs: ${r}`), [];
  }
}
async function At(e, t, o) {
  if (!c.lstatSync(o).isDirectory())
    return !1;
  const s = d.basename(o);
  if (!s.includes(e))
    return !1;
  const i = s.match(
    new RegExp(`^(.+)${e}(\\d+)-`)
  );
  if (!i)
    return !1;
  const n = {
    executableName: i[1],
    pid: i[2]
  };
  if (await Ct(n.pid, n.executableName))
    return !1;
  const l = Date.now() - t;
  return c.statSync(o).mtime.getTime() < l;
}
async function Ct(e, t) {
  const [o] = await new Promise(
    (r, s) => {
      Je.list(
        {
          pid: e,
          name: t,
          // Remove path from executable name in the results.
          clean: !0
        },
        (i, n) => {
          i ? s(i) : r(n);
        }
      );
    }
  );
  return !!o && o.pid === e && o.command === t;
}
function Dt(e) {
  return process.env.CI === "true" || process.env.CI === "1" || process.env.GITHUB_ACTIONS === "true" || process.env.GITHUB_ACTIONS === "1" || (process.env.TERM || "").toLowerCase() === "dumb" ? !1 : e ? !!e.isTTY : process.stdout.isTTY;
}
class Rt {
  constructor(t) {
    this.lastProgressLine = "", this.progressActive = !1, this.verbosity = t.verbosity, this.writeStream = t.writeStream || process.stdout;
  }
  get isTTY() {
    return !!this.writeStream.isTTY;
  }
  /**
   * Determines if progress updates should be rendered.
   *
   * Returns false when output is piped, redirected, or in CI environments.
   * This prevents progress spam in logs - users only see the final outcome.
   */
  get shouldRender() {
    return Dt(this.writeStream);
  }
  get isQuiet() {
    return this.verbosity === "quiet";
  }
  /**
   * ANSI formatting helpers.
   *
   * These only apply color codes when outputting to a terminal (TTY).
   * When piped to files or non-TTY streams, they return plain text to
   * avoid polluting logs with escape sequences.
   */
  bold(t) {
    return this.isTTY ? `\x1B[1m${t}\x1B[0m` : t;
  }
  dim(t) {
    return this.isTTY ? `\x1B[2m${t}\x1B[0m` : t;
  }
  green(t) {
    return this.isTTY ? `\x1B[32m${t}\x1B[0m` : t;
  }
  cyan(t) {
    return this.isTTY ? `\x1B[36m${t}\x1B[0m` : t;
  }
  yellow(t) {
    return this.isTTY ? `\x1B[33m${t}\x1B[0m` : t;
  }
  red(t) {
    return this.isTTY ? `\x1B[31m${t}\x1B[0m` : t;
  }
  printBanner() {
    if (this.isQuiet) return;
    const t = this.bold("WordPress Playground CLI");
    this.writeStream.write(`
${t}

`);
  }
  /**
   * Prints the configuration summary before starting the server.
   *
   * Displays PHP/WordPress versions, enabled extensions, all mounts
   * (with auto-mounts labeled), and any loaded blueprint. This gives
   * users a clear view of what's configured before the server boots.
   */
  printConfig(t) {
    if (this.isQuiet) return;
    const o = [];
    o.push(
      `${this.dim("PHP")} ${this.cyan(t.phpVersion)}  ${this.dim("WordPress")} ${this.cyan(t.wpVersion)}`
    );
    const r = [];
    if (t.intl && r.push("intl"), t.redis && r.push("redis"), t.memcached && r.push("memcached"), t.xdebug && r.push(this.yellow("xdebug")), r.length > 0 && o.push(`${this.dim("Extensions")} ${r.join(", ")}`), t.mounts.length > 0)
      for (const s of t.mounts) {
        const i = s.autoMounted ? ` ${this.dim("(auto-mount)")}` : "";
        o.push(
          `${this.dim("Mount")} ${s.hostPath} ${this.dim("→")} ${s.vfsPath}${i}`
        );
      }
    t.blueprint && o.push(`${this.dim("Blueprint")} ${t.blueprint}`), this.writeStream.write(o.join(`
`) + `

`);
  }
  /**
   * Starts a progress indicator that updates in-place.
   *
   * Subsequent updateProgress() calls rewrite the same line in TTY mode.
   * When output is piped or redirected, progress is completely skipped.
   */
  startProgress(t) {
    this.isQuiet || this.shouldRender && (this.progressActive = !0, this.updateProgress(t));
  }
  /**
   * Updates the current progress message and optional percentage.
   *
   * Rewrites the current line using ANSI cursor control in TTY mode.
   * Identical messages are skipped to prevent flickering. When piped,
   * this method does nothing (early return via shouldRender check).
   */
  updateProgress(t, o) {
    if (this.isQuiet || !this.shouldRender) return;
    this.progressActive || (this.progressActive = !0);
    let r = `${t}`;
    o !== void 0 && (r = `${t} ${this.dim(`${o}%`)}`), r !== this.lastProgressLine && (this.lastProgressLine = r, this.isTTY ? (this.writeStream.cursorTo(0), this.writeStream.write(r), this.writeStream.clearLine(1)) : this.writeStream.write(`${r}
`));
  }
  /**
   * Completes the progress indicator and moves to a new line.
   *
   * Optionally displays a final message before finishing. In TTY mode,
   * this ensures the cursor moves to the next line after the progress.
   */
  finishProgress(t) {
    this.isQuiet || this.shouldRender && (t && (this.isTTY ? (this.writeStream.cursorTo(0), this.writeStream.write(`${t}`), this.writeStream.clearLine(1)) : this.writeStream.write(`${t}
`)), this.isTTY && this.writeStream.write(`
`), this.progressActive = !1, this.lastProgressLine = "");
  }
  /**
   * Prints a status message, interrupting any active progress.
   *
   * Unlike progress updates, status messages are always printed on their
   * own line. Any active progress indicator is cleared before the message.
   */
  printStatus(t) {
    this.isQuiet || (this.progressActive && this.isTTY && (this.writeStream.cursorTo(0), this.writeStream.clearLine(0)), this.writeStream.write(`${t}
`), this.progressActive = !1, this.lastProgressLine = "");
  }
  /**
   * Prints an error message.
   *
   * Errors are always shown, even in quiet mode, and interrupt any
   * active progress display to ensure visibility.
   */
  printError(t) {
    this.progressActive && this.isTTY && (this.writeStream.cursorTo(0), this.writeStream.clearLine(0), this.progressActive = !1), this.writeStream.write(`${this.red("Error:")} ${t}
`);
  }
  /**
   * Prints the final "server ready" message with the URL.
   *
   * Note: The exact wording "WordPress is running on" is checked by
   * CI tests, so changes to this string will break test assertions.
   */
  printReady(t, o) {
    if (this.isQuiet) return;
    const r = o === 1 ? "worker" : "workers";
    this.writeStream.write(
      `
${this.green("Ready!")} WordPress is running on ${this.bold(t)} ${this.dim(`(${o} ${r})`)}

`
    );
  }
  printWarning(t) {
    this.isQuiet || this.writeStream.write(`${this.yellow("Warning:")} ${t}
`);
  }
  /**
   * Prints the phpMyAdmin URL when the --phpmyadmin flag is enabled.
   */
  printPhpMyAdminUrl(t) {
    this.isQuiet || this.writeStream.write(
      `${this.cyan("phpMyAdmin")} available at ${this.bold(t)}

`
    );
  }
}
const be = {
  Quiet: { name: "quiet", severity: F.Fatal },
  Normal: { name: "normal", severity: F.Info },
  Debug: { name: "debug", severity: F.Debug }
};
async function bo(e) {
  try {
    const t = {
      "site-url": {
        describe: "Site URL to use for WordPress. Defaults to http://127.0.0.1:{port}",
        type: "string"
      },
      php: {
        describe: "PHP version to use.",
        type: "string",
        default: M,
        choices: R
      },
      wp: {
        describe: "WordPress version to use.",
        type: "string",
        default: "latest"
      },
      define: {
        describe: 'Define PHP string constants (can be used multiple times). Format: NAME value. These constants are set via php.defineConstant() and only exist for the current request. Examples: --define API_KEY secret --define CON=ST "va=lu=e"',
        type: "string",
        nargs: 2,
        array: !0,
        coerce: dt
      },
      "define-bool": {
        describe: 'Define PHP boolean constants (can be used multiple times). Format: NAME value. Value must be "true", "false", "1", or "0". Examples: --define-bool WP_DEBUG true --define-bool MY_FEATURE false',
        type: "string",
        nargs: 2,
        array: !0,
        coerce: pt
      },
      "define-number": {
        describe: "Define PHP number constants (can be used multiple times). Format: NAME value. Examples: --define-number LIMIT 100 --define-number RATE 45.67",
        type: "string",
        nargs: 2,
        array: !0,
        coerce: ct
      },
      // @TODO: Support read-only mounts, e.g. via WORKERFS, a custom
      // ReadOnlyNODEFS, or by copying the files into MEMFS
      mount: {
        describe: "Mount a directory to the PHP runtime (can be used multiple times). Format: /host/path:/vfs/path",
        type: "array",
        string: !0,
        coerce: N
      },
      "mount-before-install": {
        describe: "Mount a directory to the PHP runtime before WordPress installation (can be used multiple times). Format: /host/path:/vfs/path",
        type: "array",
        string: !0,
        coerce: N
      },
      "mount-dir": {
        describe: 'Mount a directory to the PHP runtime (can be used multiple times). Format: "/host/path" "/vfs/path"',
        type: "array",
        nargs: 2,
        array: !0,
        coerce: ie
      },
      "mount-dir-before-install": {
        describe: 'Mount a directory before WordPress installation (can be used multiple times). Format: "/host/path" "/vfs/path"',
        type: "string",
        nargs: 2,
        array: !0,
        coerce: ie
      },
      login: {
        describe: "Should log the user in",
        type: "boolean",
        default: !1
      },
      blueprint: {
        describe: "Blueprint to execute.",
        type: "string"
      },
      "blueprint-may-read-adjacent-files": {
        describe: 'Consent flag: Allow "bundled" resources in a local blueprint to read files in the same directory as the blueprint file.',
        type: "boolean",
        default: !1
      },
      "wordpress-install-mode": {
        describe: "Control how Playground prepares WordPress before booting.",
        type: "string",
        default: "download-and-install",
        choices: [
          "download-and-install",
          "install-from-existing-files",
          "install-from-existing-files-if-needed",
          "do-not-attempt-installing"
        ]
      },
      "skip-wordpress-install": {
        describe: "[Deprecated] Use --wordpress-install-mode instead.",
        type: "boolean",
        hidden: !0
      },
      "skip-sqlite-setup": {
        describe: "Skip the SQLite integration plugin setup to allow the WordPress site to use MySQL.",
        type: "boolean",
        default: !1
      },
      // Hidden - Deprecated in favor of verbosity
      quiet: {
        describe: "Do not output logs and progress messages.",
        type: "boolean",
        default: !1,
        hidden: !0
      },
      verbosity: {
        describe: "Output logs and progress messages.",
        type: "string",
        choices: Object.values(be).map(
          (a) => a.name
        ),
        default: "normal"
      },
      debug: {
        describe: "Print PHP error log content if an error occurs during Playground boot.",
        type: "boolean",
        default: !1,
        // Hide this deprecated option. Use verbosity=debug instead.
        hidden: !0
      },
      "auto-mount": {
        describe: "Automatically mount the specified directory. If no path is provided, mount the current working directory. You can mount a WordPress directory, a plugin directory, a theme directory, a wp-content directory, or any directory containing PHP and HTML files.",
        type: "string"
      },
      "follow-symlinks": {
        describe: `Allow Playground to follow symlinks by automatically mounting symlinked directories and files encountered in mounted directories. 
Warning: Following symlinks will expose files outside mounted directories to Playground and could be a security risk.`,
        type: "boolean",
        default: !1
      },
      "experimental-trace": {
        describe: "Print detailed messages about system behavior to the console. Useful for troubleshooting.",
        type: "boolean",
        default: !1,
        // Hide this option because we want to replace with a more general log-level flag.
        hidden: !0
      },
      "internal-cookie-store": {
        describe: "Enable internal cookie handling. When enabled, Playground will manage cookies internally using an HttpCookieStore that persists cookies across requests. When disabled, cookies are handled externally (e.g., by a browser in Node.js environments).",
        type: "boolean",
        default: !1
      },
      intl: {
        describe: "Enable Intl.",
        type: "boolean",
        default: !0
      },
      redis: {
        describe: "Enable Redis (requires JSPI support).",
        type: "boolean"
        // No default - will be determined at runtime based on JSPI availability
      },
      memcached: {
        describe: "Enable Memcached.",
        type: "boolean"
        // No default - will be determined at runtime based on JSPI availability
      },
      xdebug: {
        describe: "Enable Xdebug.",
        type: "boolean",
        default: !1
      },
      "experimental-unsafe-ide-integration": {
        describe: "Enable experimental IDE development tools. This option edits IDE config files to set Xdebug path mappings and web server details. CAUTION: If there are bugs, this feature may break your IDE config files. Please consider backing up your IDE configs before using this feature.",
        type: "string",
        // The empty value means the option is enabled for all
        // supported IDEs and, if needed, will create the relevant
        // config file for each.
        choices: ["", "vscode", "phpstorm"],
        coerce: (a) => a === "" ? ["vscode", "phpstorm"] : [a]
      },
      "experimental-blueprints-v2-runner": {
        describe: "Use the experimental Blueprint V2 runner.",
        type: "boolean",
        default: !1,
        // Remove the "hidden" flag once Blueprint V2 is fully supported
        hidden: !0
      },
      mode: {
        describe: "Blueprints v2 runner mode to use. This option is required when using the --experimental-blueprints-v2-runner flag with a blueprint.",
        type: "string",
        choices: ["create-new-site", "apply-to-existing-site"],
        // Remove the "hidden" flag once Blueprint V2 is fully supported
        hidden: !0
      },
      phpmyadmin: {
        describe: "Install phpMyAdmin for database management. The phpMyAdmin URL will be printed after boot. Optionally specify a custom URL path (default: /phpmyadmin).",
        type: "string",
        coerce: (a) => a === "" ? "/phpmyadmin" : a
      }
    }, o = {
      port: {
        describe: "Port to listen on when serving. Defaults to 9400 when available.",
        type: "number"
      },
      "experimental-multi-worker": {
        deprecated: "This option is not needed. Multiple workers are always used.",
        describe: "Enable experimental multi-worker support which requires a /wordpress directory backed by a real filesystem. Pass a positive number to specify the number of workers to use. Otherwise, default to the number of CPUs minus 1.",
        type: "number"
      },
      "experimental-devtools": {
        describe: "Enable experimental browser development tools.",
        type: "boolean"
      }
    }, r = {
      path: {
        describe: "Path to the project directory. Playground will auto-detect if this is a plugin, theme, wp-content, or WordPress directory.",
        type: "string",
        default: process.cwd()
      },
      php: {
        describe: "PHP version to use.",
        type: "string",
        default: M,
        choices: R
      },
      wp: {
        describe: "WordPress version to use.",
        type: "string",
        default: "latest"
      },
      port: {
        describe: "Port to listen on. Defaults to 9400 when available.",
        type: "number"
      },
      blueprint: {
        describe: "Path to a Blueprint JSON file to execute on startup.",
        type: "string"
      },
      login: {
        describe: "Auto-login as the admin user.",
        type: "boolean",
        default: !0
      },
      xdebug: {
        describe: "Enable Xdebug for debugging.",
        type: "boolean",
        default: !1
      },
      "experimental-unsafe-ide-integration": t["experimental-unsafe-ide-integration"],
      "skip-browser": {
        describe: "Do not open the site in your default browser on startup.",
        type: "boolean",
        default: !1
      },
      quiet: {
        describe: "Suppress non-essential output.",
        type: "boolean",
        default: !1
      },
      // Advanced options for power users who need more control
      "site-url": {
        describe: "Override the site URL. By default, derived from the port (http://127.0.0.1:<port>).",
        type: "string"
      },
      mount: {
        describe: "Mount a directory to the PHP runtime (can be used multiple times). Format: /host/path:/vfs/path. Use this for additional mounts beyond auto-detection.",
        type: "array",
        string: !0,
        coerce: N
      },
      reset: {
        describe: "Deletes the stored site directory and starts a new site from scratch.",
        type: "boolean",
        default: !1
      },
      "no-auto-mount": {
        describe: "Disable automatic project type detection. Use --mount to manually specify mounts instead.",
        type: "boolean",
        default: !1
      },
      // Define constants
      define: t.define,
      "define-bool": t["define-bool"],
      "define-number": t["define-number"],
      // Tools
      phpmyadmin: t.phpmyadmin
    }, s = {
      outfile: {
        describe: "When building, write to this output file.",
        type: "string",
        default: "wordpress.zip"
      }
    }, i = Ne(e).usage("Usage: wp-playground <command> [options]").command(
      "start",
      "Start a local WordPress server with automatic project detection (recommended)",
      (a) => a.usage(
        `Usage: wp-playground start [options]

The easiest way to run WordPress locally. Automatically detects
if your directory contains a plugin, theme, wp-content, or
WordPress installation and configures everything for you.

Examples:
  wp-playground start                    # Start in current directory
  wp-playground start --path=./my-plugin # Start with a specific path
  wp-playground start --wp=6.7 --php=8.3 # Use specific versions
  wp-playground start --skip-browser     # Skip opening browser
  wp-playground start --no-auto-mount    # Disable auto-detection`
      ).options(r)
    ).command(
      "server",
      "Start a local WordPress server (advanced, low-level)",
      (a) => a.options({
        ...t,
        ...o
      })
    ).command(
      "run-blueprint",
      "Execute a Blueprint without starting a server",
      (a) => a.options({ ...t })
    ).command(
      "build-snapshot",
      "Build a ZIP snapshot of a WordPress site based on a Blueprint",
      (a) => a.options({
        ...t,
        ...s
      })
    ).demandCommand(1, "Please specify a command").strictCommands().conflicts(
      "experimental-unsafe-ide-integration",
      "experimental-devtools"
    ).showHelpOnFail(!1).fail((a, m, y) => {
      if (m)
        throw m;
      a && a.includes("Please specify a command") && (y.showHelp(), console.error(`
` + a), process.exit(1)), console.error(a), process.exit(1);
    }).strictOptions().check(async (a) => {
      if (a["skip-wordpress-install"] === !0 && (a["wordpress-install-mode"] = "do-not-attempt-installing", a.wordpressInstallMode = "do-not-attempt-installing"), a.wp !== void 0 && typeof a.wp == "string" && !vt(a.wp))
        try {
          new URL(a.wp);
        } catch {
          throw new Error(
            'Unrecognized WordPress version. Please use "latest", a URL, or a numeric version such as "6.2", "6.0.1", "6.2-beta1", or "6.2-RC1"'
          );
        }
      const m = a["site-url"];
      if (typeof m == "string" && m.trim() !== "")
        try {
          new URL(m);
        } catch {
          throw new Error(
            `Invalid site-url "${m}". Please provide a valid URL (e.g., http://localhost:8080 or https://example.com)`
          );
        }
      if (a["auto-mount"]) {
        let y = !1;
        try {
          y = c.statSync(
            a["auto-mount"]
          ).isDirectory();
        } catch {
          y = !1;
        }
        if (!y)
          throw new Error(
            `The specified --auto-mount path is not a directory: '${a["auto-mount"]}'.`
          );
      }
      if (a["experimental-blueprints-v2-runner"] === !0)
        throw new Error(
          "Blueprints v2 are temporarily disabled while we rework their runtime implementation."
        );
      if (a.mode !== void 0)
        throw new Error(
          "The --mode option requires the --experimentalBlueprintsV2Runner flag."
        );
      return !0;
    });
    i.wrap(i.terminalWidth());
    const n = await i.argv, l = n._[0];
    ["start", "run-blueprint", "server", "build-snapshot"].includes(
      l
    ) || (i.showHelp(), process.exit(1));
    const u = n.define || {};
    !("WP_DEBUG" in u) && !("WP_DEBUG_LOG" in u) && !("WP_DEBUG_DISPLAY" in u) && (u.WP_DEBUG = "true", u.WP_DEBUG_LOG = "true", u.WP_DEBUG_DISPLAY = "true");
    const w = {
      ...n,
      define: u,
      command: l,
      mount: [
        ...n.mount || [],
        ...n["mount-dir"] || []
      ],
      "mount-before-install": [
        ...n["mount-before-install"] || [],
        ...n["mount-dir-before-install"] || []
      ]
    }, S = await Lt(w);
    S === void 0 && process.exit(0);
    const p = /* @__PURE__ */ (() => {
      let a;
      return async () => {
        a === void 0 && (a = S[Symbol.asyncDispose]()), await a, process.exit(0);
      };
    })();
    return process.on("SIGINT", p), process.on("SIGTERM", p), {
      [Symbol.asyncDispose]: async () => {
        process.off("SIGINT", p), process.off("SIGTERM", p), await S[Symbol.asyncDispose]();
      },
      [ge]: { cliServer: S }
    };
  } catch (t) {
    if (console.error(t), !(t instanceof Error))
      throw t;
    if (process.argv.includes("--debug"))
      $e(t);
    else {
      const r = [];
      let s = t;
      do
        r.push(s.message), s = s.cause;
      while (s instanceof Error);
      console.error(
        "\x1B[1m" + r.join(" caused by: ") + "\x1B[0m"
      );
    }
    process.exit(1);
  }
}
function ae(e, t) {
  return e.find(
    (o) => o.vfsPath.replace(/\/$/, "") === t.replace(/\/$/, "")
  );
}
const ge = Symbol("playground-cli-testing"), I = (e) => process.stdout.isTTY ? "\x1B[1m" + e + "\x1B[0m" : e, Wt = (e) => process.stdout.isTTY ? "\x1B[31m" + e + "\x1B[0m" : e, le = (e) => process.stdout.isTTY ? `\x1B[2m${e}\x1B[0m` : e, C = (e) => process.stdout.isTTY ? `\x1B[3m${e}\x1B[0m` : e, O = (e) => process.stdout.isTTY ? `\x1B[33m${e}\x1B[0m` : e;
async function Lt(e) {
  let t;
  const o = e.internalCookieStore ? new Ee() : void 0, r = [], s = /* @__PURE__ */ new Map();
  if (e.command === "start" && (e = Ut(e)), e.autoMount !== void 0 && (e.autoMount === "" && (e = { ...e, autoMount: process.cwd() }), e = he(e)), e.wordpressInstallMode === void 0 && (e.wordpressInstallMode = "download-and-install"), e.quiet && (e.verbosity = "quiet", delete e.quiet), e.debug && (e.verbosity = "debug", delete e.debug), e.verbosity) {
    const p = Object.values(be).find(
      (a) => a.name === e.verbosity
    ).severity;
    P.setSeverityFilterLevel(p);
  }
  if (e.intl || (e.intl = !0), e.redis === void 0 && (e.redis = await se()), e.memcached === void 0 && (e.memcached = await se()), e.phpmyadmin) {
    if (e.phpmyadmin === !0 && (e.phpmyadmin = "/phpmyadmin"), e.skipSqliteSetup)
      throw new Error(
        "--phpmyadmin requires SQLite. Cannot be used with --skip-sqlite-setup."
      );
    e.pathAliases = [
      {
        urlPrefix: e.phpmyadmin,
        fsPath: re
      }
    ];
  }
  const i = new Rt({
    verbosity: e.verbosity || "normal"
  });
  e.command === "server" && (i.printBanner(), i.printConfig({
    phpVersion: e.php || M,
    wpVersion: e.wp || "latest",
    port: e.port ?? 9400,
    xdebug: !!e.xdebug,
    intl: !!e.intl,
    redis: !!e.redis,
    memcached: !!e.memcached,
    mounts: [
      ...e.mount || [],
      ...e["mount-before-install"] || []
    ],
    blueprint: typeof e.blueprint == "string" ? e.blueprint : void 0
  }));
  const n = e.command === "server" ? e.port ?? 9400 : 0, l = new Te();
  let u = !1, w = !0;
  const S = await wt({
    port: e.port ? e.port : await ht(n) ? 0 : n,
    onBind: async (p, a) => {
      const m = "127.0.0.1", y = `http://${m}:${a}`, W = e["site-url"] || y, L = 6, z = "-playground-cli-site-", E = await kt(z);
      P.debug(`Native temp dir for VFS root: ${E.path}`);
      const B = "WP Playground CLI - Listen for Xdebug", G = ".playground-xdebug-root", U = d.join(process.cwd(), G);
      if (await Ke(U), e.xdebug) {
        const f = {
          hostPath: d.join(".", d.sep, G),
          vfsPath: "/"
        };
        if (R.indexOf(
          e.php || M
        ) <= R.indexOf("8.5"))
          await oe(
            E.path,
            U,
            process.platform
          ), e.xdebug = et({
            cwd: process.cwd(),
            mounts: [
              f,
              ...e["mount-before-install"] || [],
              ...e.mount || []
            ],
            pathSkippings: [
              "/dev/",
              "/home/",
              "/internal/",
              "/request/",
              "/proc/"
            ]
          }), console.log(I("Xdebug configured successfully")), console.log(
            O("Playground source root: ") + ".playground-xdebug-root" + C(
              le(
                " – you can set breakpoints and preview Playground's VFS structure in there."
              )
            )
          );
        else if (e.experimentalUnsafeIdeIntegration) {
          await oe(
            E.path,
            U,
            process.platform
          );
          try {
            await tt(
              B,
              process.cwd()
            );
            const h = typeof e.xdebug == "object" ? e.xdebug : {}, b = await ot({
              name: B,
              host: m,
              port: a,
              ides: e.experimentalUnsafeIdeIntegration,
              cwd: process.cwd(),
              mounts: [
                f,
                ...e["mount-before-install"] || [],
                ...e.mount || []
              ],
              ideKey: h.ideKey || "WPPLAYGROUNDCLI"
            }), v = e.experimentalUnsafeIdeIntegration, x = v.includes("vscode"), T = v.includes("phpstorm"), X = Object.values(b);
            console.log(""), X.length > 0 ? (console.log(
              I("Xdebug configured successfully")
            ), console.log(
              O("Updated IDE config: ") + X.join(" ")
            ), console.log(
              O("Playground source root: ") + ".playground-xdebug-root" + C(
                le(
                  " – you can set breakpoints and preview Playground's VFS structure in there."
                )
              )
            )) : (console.log(
              I("Xdebug configuration failed.")
            ), console.log(
              "No IDE-specific project settings directory was found in the current working directory."
            )), console.log(""), x && b.vscode && (console.log(
              I("VS Code / Cursor instructions:")
            ), console.log(
              "  1. Ensure you have installed an IDE extension for PHP Debugging"
            ), console.log(
              `     (The ${I("PHP Debug")} extension by ${I(
                "Xdebug"
              )} has been a solid option)`
            ), console.log(
              "  2. Open the Run and Debug panel on the left sidebar"
            ), console.log(
              `  3. Select "${C(
                B
              )}" from the dropdown`
            ), console.log('  3. Click "start debugging"'), console.log(
              "  5. Set a breakpoint. For example, in .playground-xdebug-root/wordpress/index.php"
            ), console.log(
              "  6. Visit Playground in your browser to hit the breakpoint"
            ), T && console.log("")), T && b.phpstorm && (console.log(I("PhpStorm instructions:")), console.log(
              `  1. Choose "${C(
                B
              )}" debug configuration in the toolbar`
            ), console.log(
              "  2. Click the debug button (bug icon)`"
            ), console.log(
              "  3. Set a breakpoint. For example, in .playground-xdebug-root/wordpress/index.php"
            ), console.log(
              "  4. Visit Playground in your browser to hit the breakpoint"
            )), console.log("");
          } catch (h) {
            throw new Error("Could not configure Xdebug", {
              cause: h
            });
          }
        }
      }
      const Pe = d.dirname(E.path), ve = 2 * 24 * 60 * 60 * 1e3;
      Mt(
        z,
        ve,
        Pe
      );
      const Q = d.join(E.path, "internal");
      j(Q);
      const Se = [
        "wordpress",
        "tools",
        // Note: These dirs are from Emscripten's "default dirs" list:
        // https://github.com/emscripten-core/emscripten/blob/f431ec220e472e1f8d3db6b52fe23fb377facf30/src/lib/libfs.js#L1400-L1402
        //
        // Any Playground process with multiple workers may assume
        // these are part of a shared filesystem, so let's recognize
        // them explicitly here.
        "tmp",
        "home"
      ];
      for (const f of Se) {
        const g = (b) => b.vfsPath === `/${f}`;
        if (!(e["mount-before-install"]?.some(g) || e.mount?.some(g))) {
          const b = d.join(
            E.path,
            f
          );
          j(b), e["mount-before-install"] === void 0 && (e["mount-before-install"] = []), e["mount-before-install"].unshift({
            vfsPath: `/${f}`,
            hostPath: b
          });
        }
      }
      if (e["mount-before-install"])
        for (const f of e["mount-before-install"])
          P.debug(
            `Mount before WP install: ${f.vfsPath} -> ${f.hostPath}`
          );
      if (e.mount)
        for (const f of e.mount)
          P.debug(
            `Mount after WP install: ${f.vfsPath} -> ${f.hostPath}`
          );
      let k;
      e["experimental-blueprints-v2-runner"] ? k = new xt(e, {
        siteUrl: W,
        cliOutput: i
      }) : (k = new It(e, {
        siteUrl: W,
        cliOutput: i
      }), typeof e.blueprint == "string" && (e.blueprint = await St({
        sourceString: e.blueprint,
        blueprintMayReadAdjacentFiles: e["blueprint-may-read-adjacent-files"] === !0
      })));
      let H = !1;
      const A = async function() {
        H || (H = !0, await Promise.all(
          r.map(async (g) => {
            await s.get(g)?.dispose(), await g.worker.terminate();
          })
        ), p && await new Promise((g) => {
          p.close(g), p.closeAllConnections();
        }), await E.cleanup());
      };
      try {
        const f = [], g = k.getWorkerType();
        for (let h = 0; h < L; h++) {
          const b = Ht(g, {
            onExit: (v) => {
              H || v === 0 && P.error(
                `Worker ${h} exited with code ${v}
`
              );
            }
          }).then(
            async (v) => {
              r.push(v);
              const x = await Ft(l), T = await k.bootRequestHandler({
                worker: v,
                fileLockManagerPort: x,
                nativeInternalDirPath: Q
              });
              return s.set(
                v,
                T
              ), [v, T];
            }
          );
          f.push(b), h === 0 && await b;
        }
        await Promise.all(f), t = Ie(
          r.map(
            (h) => s.get(h)
          )
        );
        {
          const h = new pe(), b = h.port1, v = h.port2;
          if (await ke(
            {
              applyPostInstallMountsToAllWorkers: async () => {
                await Promise.all(
                  Array.from(
                    s.values()
                  ).map(
                    (x) => x.mountAfterWordPressInstall(
                      e.mount || []
                    )
                  )
                );
              }
            },
            void 0,
            b
          ), await k.bootWordPress(
            t,
            v
          ), b.close(), u = !0, !e["experimental-blueprints-v2-runner"]) {
            const x = await k.compileInputBlueprint(
              e["additional-blueprint-steps"] || []
            );
            x && await ee(
              x,
              t
            );
          }
          if (e.phpmyadmin && !await t.fileExists(
            `${re}/index.php`
          )) {
            const x = await st(), T = await de({ steps: x });
            await ee(T, t);
          }
          if (e.command === "build-snapshot") {
            await Ot(t, e.outfile), i.printStatus(`Exported to ${e.outfile}`), await A();
            return;
          } else if (e.command === "run-blueprint") {
            i.finishProgress("Done"), await A();
            return;
          }
        }
        if (i.finishProgress(), i.printReady(y, L), e.phpmyadmin) {
          const h = d.join(
            e.phpmyadmin,
            it
          );
          i.printPhpMyAdminUrl(
            new URL(h, y).toString()
          );
        }
        return e.xdebug && e.experimentalDevtools && (await Qe({
          phpInstance: t,
          phpRoot: "/wordpress"
        })).start(), {
          playground: t,
          server: p,
          serverUrl: y,
          [Symbol.asyncDispose]: A,
          [ge]: {
            workerThreadCount: L
          }
        };
      } catch (f) {
        if (e.verbosity !== "debug")
          throw f;
        let g = "";
        throw await t?.fileExists(Z) && (g = await t.readFileAsText(Z)), await A(), new Error(g, { cause: f });
      }
    },
    async handleRequest(p) {
      if (!u)
        return J.forHttpCode(
          502,
          "WordPress is not ready yet"
        );
      if (w) {
        w = !1;
        const m = {
          "Content-Type": ["text/plain"],
          "Content-Length": ["0"],
          Location: [p.url]
        };
        return p.headers?.cookie?.includes(
          "playground_auto_login_already_happened"
        ) && (m["Set-Cookie"] = [
          "playground_auto_login_already_happened=1; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Path=/"
        ]), J.fromPHPResponse(
          new Me(302, m, new Uint8Array())
        );
      }
      o && (p = {
        ...p,
        headers: {
          ...p.headers,
          // While we have an internal cookie store, we
          // completely replace the incoming request's Cookie
          // header with the cookies from our store. This avoids
          // getting into a strange state where both browser and
          // server are managing cookies.
          cookie: o.getCookieRequestHeader()
        }
      });
      const a = await t.requestStreamed(p);
      if (o) {
        const m = await a.headers;
        o.rememberCookiesFromResponseHeaders(m), delete m["set-cookie"];
      }
      return a;
    }
  }).catch((p) => {
    i.printError(p.message), process.exit(1);
  });
  return S && e.command === "start" && !e.skipBrowser && Nt(S.serverUrl), S;
}
function Ut(e) {
  let t = { ...e, command: "server" };
  e.noAutoMount || (t.autoMount = d.resolve(process.cwd(), t.path ?? ""), t = he(t), delete t.autoMount);
  const o = ae(
    t["mount-before-install"] || [],
    "/wordpress"
  ) || ae(t.mount || [], "/wordpress");
  if (o)
    console.log("Site files stored at:", o?.hostPath), e.reset && (console.log(""), console.log(
      Wt(
        "This site is not managed by Playground CLI and cannot be reset."
      )
    ), console.log(
      "(It's not stored in the ~/.wordpress-playground/sites/<site-id> directory.)"
    ), console.log(""), console.log(
      "You may still remove the site's directory manually if you wish."
    ), process.exit(1));
  else {
    const r = t.autoMount || process.cwd(), s = rt("sha256").update(r).digest("hex"), i = q.homedir(), n = d.join(
      i,
      ".wordpress-playground/sites",
      s
    );
    console.log("Site files stored at:", n), Y(n) && e.reset && (console.log("Resetting site..."), Re(n, { recursive: !0 })), j(n, { recursive: !0 }), t["mount-before-install"] = [
      ...t["mount-before-install"] || [],
      { vfsPath: "/wordpress", hostPath: n }
    ], t.wordpressInstallMode = We(n).length === 0 ? (
      // Only download WordPress on the first run when the site directory is still
      // empty.
      "download-and-install"
    ) : (
      // After that, reuse the WordPress installation from the initial run.
      "install-from-existing-files-if-needed"
    );
  }
  return t;
}
const _ = new xe();
function Ht(e, { onExit: t } = {}) {
  let o;
  return e === "v1" ? o = new te(new URL("./worker-thread-v1.js", import.meta.url)) : o = new te(new URL("./worker-thread-v2.js", import.meta.url)), new Promise((r, s) => {
    const i = _.claim();
    o.once("message", function(l) {
      l.command === "worker-script-initialized" && r({
        processId: i,
        worker: o,
        phpPort: l.phpPort
      });
    }), o.once("error", function(l) {
      _.release(i), console.error(l);
      const u = new Error(
        `Worker failed to load worker. ${l.message ? `Original error: ${l.message}` : ""}`
      );
      s(u);
    });
    let n = !1;
    o.once("spawn", () => {
      n = !0;
    }), o.once("exit", (l) => {
      _.release(i), n || s(new Error(`Worker exited before spawning: ${l}`)), t?.(l);
    });
  });
}
async function Ft(e) {
  const { port1: t, port2: o } = new pe();
  return await Be(e, t), o;
}
function Nt(e) {
  const t = q.platform();
  let o;
  switch (t) {
    case "darwin":
      o = `open "${e}"`;
      break;
    case "win32":
      o = `start "" "${e}"`;
      break;
    default:
      o = `xdg-open "${e}"`;
      break;
  }
  me(o, (r) => {
    r && P.debug(`Could not open browser: ${r.message}`);
  });
}
async function Ot(e, t) {
  await e.run({
    code: `<?php
		$zip = new ZipArchive();
		if(false === $zip->open('/tmp/build.zip', ZipArchive::CREATE | ZipArchive::OVERWRITE)) {
			throw new Exception('Failed to create ZIP');
		}
		$files = new RecursiveIteratorIterator(
			new RecursiveDirectoryIterator('/wordpress')
		);
		foreach ($files as $file) {
			echo $file . PHP_EOL;
			if (!$file->isFile()) {
				continue;
			}
			$zip->addFile($file->getPathname(), $file->getPathname());
		}
		$zip->close();

	`
  });
  const o = await e.readFileAsBuffer("/tmp/build.zip");
  c.writeFileSync(t, o);
}
export {
  be as L,
  yo as a,
  Dt as b,
  ge as i,
  we as m,
  bo as p,
  Lt as r,
  Ht as s
};
